File: show.erl
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% 
% this module contains funs for showing simulation
%
% Ian Tsybulkin, 2016
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

-module(show).
-export([make_demo/2
		]).
-define(PI,3.1416).
-define(Xt, 0.05).
-define(Yt, 0).

make_demo(Log,Tau) ->
	{ok,Out} = file:open("jumper.html", [write]),
	io:format(Out, "<svg width='1000' height='650'>~n",[]),
	
	io:format(Out,"~n<polygon points='~w,~w ~w,~w ~w,~w ~w,~w' 
					style='fill:brown;stroke:purple;stroke-width:1' >~n",[0,0,0,0,0,0,0,0]),
	draw_leg(Log,0,Tau,Out),

	io:format(Out,"\t<line x1='~w' y1='~w' x2='~w' y2='~w' 
					style='fill:brown;stroke:purple;stroke-width:15' >~n",[0,0,0,0]),
	draw_body(Log,0,Tau,Out),
	io:format(Out, "</svg>\n",[]),
	file:close(Out).


draw_leg([{_,_,B,_,_}|Log],T,Tau,Out) ->
	{Xt1,Yt1}=px({?Xt,?Yt}),
	{Xa,Ya} = px(phy:get_XYa(?Xt,?Yt,B)),
	{Xh,Yh} = px(phy:get_XYh(?Xt,?Yt,B)),
	{X1,Y1} = px(phy:get_XY1(?Xt,?Yt,B)),
	io:format(Out,"\t<set attributeName='points' attributeType='XML'
		to='~w,~w ~w,~w ~w,~w ~w,~w' begin='~ws'  />~n",
         	[Xt1,Yt1,Xa,Ya,Xh,Yh,X1,Y1,T]),
	draw_leg(Log,T+Tau,Tau,Out);
draw_leg([],_,_,Out) ->
	io:format(Out,"</polygon>~n",[]).


draw_body([{A,_,B,_,_}|Log],T,Tau,Out) ->
	{X1,Y1} = px(phy:get_XYh(?Xt,?Yt,B)),
	{X2,Y2} = px(phy:get_XY2(?Xt,?Yt,A,B)),
	
	io:format(Out,"<set attributeName='x1' attributeType='XML'
		to='~w' begin='~w'  />~n",[X1,T]),
	io:format(Out,"<set attributeName='y1' attributeType='XML'
		to='~w' begin='~w'  />~n",[Y1,T]),
	io:format(Out,"<set attributeName='x2' attributeType='XML'
		to='~w' begin='~w'  />~n",[X2,T]),
	io:format(Out,"<set attributeName='y2' attributeType='XML'
		to='~w' begin='~w'  />~n",[Y2,T]),
		
	draw_body(Log,T+Tau,Tau,Out);
draw_body([],_,_,Out) ->
	io:format(Out,"</line>~n",[]).




px({X,Y}) -> {300+round(X*2000), 400-round(Y*2000)}.

degr(Rad) -> round(180*Rad/?PI).

File: show.py
from params import *
from numpy import sin, cos

x,y = 0.3, 0 # coords of the toe tip


def show(anim_file, state_log,tau):
	f = open(anim_file,'w')
	f.write("<html>\n<body>\n<svg width='1000' height='650'>\n")

	draw_line(f,[((x,y),(x+L3*cos(b),y+L3*sin(b))) for (x,y,_,b,_) in state_log ],5,'blue',tau)
	draw_line(f,[((x+L3*cos(b),y+L3*sin(b)),(x+L3*cos(b)+L2*cos(b+g),y+L3*sin(b)+L2*sin(b+g))) 
		for (x,y,_,b,g) in state_log ],7,'rgb(50,50,200)',tau)
	draw_line(f,[((x+L3*cos(b)+L2*cos(b+g),y+L3*sin(b)+L2*sin(b+g)),
			(x+L3*cos(b)+L2*cos(b+g)+2*L1*cos(a+b+g),y+L3*sin(b)+L2*sin(b+g)+2*L1*sin(a+b+g))) 
		for (x,y,a,b,g) in state_log ],9,'rgb(70,70,150)',tau)

	f.write("</svg>\n</body>\n</html>")
	f.close()


def draw_line(f, points_list, width, color, tau):
	f.write("\t<line x1='%i' y1='%i' x2='%i' y2='%i' \
		style='stroke:%s;stroke-width:%i' >\n" % (0,0,0,0,color,width))
	T = 0
	for xy1,xy2 in points_list:
		x1,y1 = px(xy1)
		x2,y2 = px(xy2)
		f.write("\t\t<set attributeName='x1' attributeType='XML'\n \
         to='%i' begin='%.2fs'  />\n" %(x1,T) ),
		f.write("\t\t<set attributeName='y1' attributeType='XML'\n \
         to='%i' begin='%.2fs'  />\n" %(y1,T) ),
		f.write("\t\t<set attributeName='x2' attributeType='XML'\n \
         to='%i' begin='%.2fs'  />\n" %(x2,T) ),
		f.write("\t\t<set attributeName='y2' attributeType='XML'\n \
         to='%i' begin='%.2fs'  />\n" %(y2,T) ),
		T += tau
	f.write("\t</line>\n")	


def px(xy):
	return ( 100+round(xy[0]*1500), 600-round(xy[1]*1500) )



